In [ ]:
pip install yfinance

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
pip install ta

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for ta: filename=ta-0.10.2-py3-none-any.whl size=29103 sha256=164ddb8c954fa27dbc64285f22b84e44a3f862e2159dbedb7da59a6f45c7daae
  Stored in directory: /root/.cache/pip/wheels/6a/76/03/7f785aaa50b9c6ec7e3fd105a62c1b2c45a034512d51e024a9
Successfully built ta


In [ ]:
import yfinance as yf
import pandas as pd
import ta
from datetime import datetime, timedelta
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_regression
import warnings

warnings.filterwarnings('ignore', category=RuntimeWarning, message='invalid value encountered in.*')


def get_stock_data(ticker, years=6):
    end_date = datetime.today()
    start_date = end_date - timedelta(days=365*years)

    stock_data = yf.download(ticker, start=start_date, end=end_date)
    return stock_data

def add_technical_indicators(data):
    # Add Stochastic Oscillator (%K)
    stoch = ta.momentum.StochasticOscillator(data['High'], data['Low'], data['Close'])
    data['%K'] = stoch.stoch()

    # Add Relative Strength Index (RSI)
    rsi = ta.momentum.RSIIndicator(data['Close'])
    data['RSI'] = rsi.rsi()

    # Add Simple Moving Average (SMA)
    sma = ta.trend.SMAIndicator(data['Close'], window=20)
    data['SMA'] = sma.sma_indicator()

    # Add Exponential Moving Average (EMA)
    ema = ta.trend.EMAIndicator(data['Close'], window=20)
    data['EMA'] = ema.ema_indicator()

    # Add Bollinger Bands
    bollinger = ta.volatility.BollingerBands(data['Close'])
    data['Bollinger_High'] = bollinger.bollinger_hband()
    data['Bollinger_Low'] = bollinger.bollinger_lband(

    # Add Average Directional Index (ADX)
    adx = ta.trend.ADXIndicator(data['High'], data['Low'], data['Close'])
    data['ADX'] = adx.adx()

    # Add On Balance Volume (OBV)
    obv = ta.volume.OnBalanceVolumeIndicator(data['Close'], data['Volume'])
    data['OBV'] = obv.on_balance_volume()

    # Add Chaikin Money Flow (CMF)
    cmf = ta.volume.ChaikinMoneyFlowIndicator(data['High'], data['Low'], data['Close'], data['Volume'])
    data['CMF'] = cmf.chaikin_money_flow()

    return data

In [ ]:
ticker = "TSLA"  # Tesla stock symbol
raw_data = get_stock_data(ticker)
processed_data = add_technical_indicators(raw_data)

processed_data.head(50)


[*********************100%***********************]  1 of 1 completed


Open       High        Low      Close  Adj Close     Volume  \
Date                                                                           
2017-04-11  20.892000  20.898001  20.366667  20.580667  20.580667   85869000   
2017-04-12  20.422667  20.563334  19.754667  19.789333  19.789333   90760500   
2017-04-13  19.780001  20.492666  19.686666  20.266666  20.266666  139269000   
2017-04-17  20.180000  20.266666  19.912001  20.096001  20.096001   62080500   
2017-04-18  19.980000  20.056000  19.860001  20.016666  20.016666   45535500   
2017-04-19  20.164000  20.441334  20.140667  20.368000  20.368000   58470000   
2017-04-20  20.434000  20.610001  20.015333  20.167334  20.167334   92241000   
2017-04-21  20.133333  20.426666  20.028000  20.373333  20.373333   67647000   
2017-04-24  20.614668  20.703333  20.401333  20.535334  20.535334   76252500   
2017-04-25  20.533333  20.931999  20.390667  20.919333  20.919333  101065500   
2017-04-26  20.824667  20.966667  20.600000  20.677999  20.677999   70425000   
2017-04-27  20.779333  20.872667  20.500000  20.575333  20.575333   52029000   
2017-04-28  20.655333  20.986668  20.533333  20.938000  20.938000   67582500   
2017-05-01  20.992001  21.816668  20.987333  21.521999  21.521999  132444000   
2017-05-02  21.600000  21.844000  21.104000  21.259333  21.259333   80742000   
2017-05-03  21.177999  21.435333  20.696667  20.734667  20.734667  107001000   
2017-05-04  20.496000  20.518000  19.384001  19.697332  19.697332  212280000   
2017-05-05  19.866667  20.570000  19.786667  20.556667  20.556667  122659500   
2017-05-08  20.726667  20.919333  20.388000  20.479334  20.479334  105097500   
2017-05-09  20.625334  21.466000  20.606667  21.417334  21.417334  145147500   
2017-05-10  21.437332  21.700001  21.208000  21.681334  21.681334   86124000   
2017-05-11  21.559999  21.733334  21.306667  21.540001  21.540001   71307000   
2017-05-12  21.698668  21.799999  21.435333  21.653999  21.653999   61824000   
2017-05-15  21.225332  21.346666  20.835333  21.058666  21.058666  114330000   
2017-05-16  21.172667  21.337334  21.009333  21.134001  21.134001   62287500   
2017-05-17  20.959333  20.975332  20.366667  20.407333  20.407333  100678500   
2017-05-18  20.466667  20.929333  20.354000  20.870667  20.870667   84807000   
2017-05-19  21.033333  21.100000  20.680000  20.722000  20.722000   70314000   
2017-05-22  20.853333  20.958000  20.453333  20.690001  20.690001   64938000   
2017-05-23  20.697332  20.715334  20.232000  20.257334  20.257334   64776000   
2017-05-24  20.434000  20.733334  20.360001  20.681334  20.681334   75499500   
2017-05-25  20.734667  21.131332  20.520666  21.122000  21.122000   75210000   
2017-05-26  21.152000  21.699333  21.087334  21.676001  21.676001  117033000   
2017-05-30  21.733334  22.418667  21.717333  22.340000  22.340000  116743500   
2017-05-31  22.512667  22.859333  22.344000  22.733999  22.733999  149451000   
2017-06-01  22.933332  22.992001  22.486000  22.691334  22.691334  114120000   
2017-06-02  22.651333  22.858667  22.395332  22.656668  22.656668   83853000   
2017-06-05  22.566668  23.229334  22.280666  23.154667  23.154667  101766000   
2017-06-06  22.980000  23.966000  22.664667  23.523333  23.523333  166302000   
2017-06-07  23.756001  24.033333  23.676001  23.976667  23.976667  140970000   
2017-06-08  24.250000  24.793333  24.014668  24.666668  24.666668  135922500   
2017-06-09  24.961332  25.124666  23.653334  23.821333  23.821333  258921000   
2017-06-12  23.865999  24.299999  23.374666  23.934000  23.934000  157765500   
2017-06-13  24.507999  25.066668  24.440666  25.063334  25.063334  177118500   
2017-06-14  25.406000  25.616667  25.087334  25.377333  25.377333  192276000   
2017-06-15  24.833332  25.030666  24.432667  25.022667  25.022667  156397500   
2017-06-16  25.198668  25.200666  24.673332  24.760000  24.760000  100965000   
2017-06-19  25.000000  25.113333  24.520000  24.653334  24.653334   98239500   
2017-06-20  25.11

In [ ]:
def create_target_and_preprocess(data, shift=1):
    data['Future_Close'] = data['Close'].shift(-shift)
    data['Trend'] = (data['Future_Close'] > data['Close']).astype(int)
    data = data.drop(['Future_Close'], axis=1)
    data = data.dropna()
    return data

def prepare_data_for_feature_selection(data, target_col):
    data = data.copy()
    X = data.drop([target_col], axis=1)
    y = data[target_col]
    return X, y

In [ ]:
ticker = "TSLA"  # Tesla stock symbol
raw_data = get_stock_data(ticker)
processed_data = add_technical_indicators(raw_data)
processed_data = create_target_and_preprocess(processed_data)
processed_data = processed_data.dropna()

[*********************100%***********************]  1 of 1 completed


In [ ]:
processed_data.head()

Open       High        Low      Close  Adj Close     Volume  \
Date                                                                           
2017-05-09  20.625334  21.466000  20.606667  21.417334  21.417334  145147500   
2017-05-10  21.437332  21.700001  21.208000  21.681334  21.681334   86124000   
2017-05-11  21.559999  21.733334  21.306667  21.540001  21.540001   71307000   
2017-05-12  21.698668  21.799999  21.435333  21.653999  21.653999   61824000   
2017-05-15  21.225332  21.346666  20.835333  21.058666  21.058666  114330000   

                   %K        RSI        SMA        EMA  Bollinger_High  \
Date                                                                     
2017-05-09  82.655837  58.706452  20.548533  20.655660       21.510998   
2017-05-10  93.387545  60.977059  20.603567  20.753343       21.685539   
2017-05-11  87.642315  59.103379  20.691100  20.828263       21.778668   
2017-05-12  92.276398  60.166585  20.760467  20.906905       21.906312   
2017-05-15  68.075857  52.491781  20.808600  20.921358       21.919085   

            Bollinger_Low  ADX        OBV       CMF  Trend  
Date                                                        
2017-05-09      19.586068  0.0   78214500  0.038443      1  
2017-05-10      19.521594  0.0  164338500  0.088723      0  
2017-05-11      19.603532  0.0   93031500  0.136940      1  
2017-05-12      19.614621  0.0  154855500  0.115903      0  
2017-05-15      19.698115  0.0   40525500  0.103686      1

##NEW IMPLEMENTATION - Running AutoML

In [ ]:
#Install the pycaret library, which is an AutoML library in Python.

!pip install pycaret

from pycaret.classification import *

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 481.8/481.8 KB 9.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 298.0/298.0 KB 35.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.1/16.1 MB 75.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 78.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 150.0/150.0 KB 14.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.8/43.8 KB 6.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 212.2/212.2 KB 28.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.4/46.4 KB 6.8 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.9/79.9 MB 12.6 MB/s eta 0:00:00


In [ ]:
# Initialize pycaret
exp_clf = setup(data=processed_data, target='Trend', train_size=0.8, session_id=42, preprocess=False)

# Train and compare models
best_model = compare_models(sort='Accuracy', n_select=1)

# Finalize the model
final_model = finalize_model(best_model)

# Make predictions on the test set
test_data = processed_data.iloc[-100:]  # Assuming you want to use the last 100 rows for testing
predictions = predict_model(final_model, data=test_data)

Processing:   0%|          | 0/65 [00:00<?, ?it/s]

#The first result (91% accuracy) is based on the validation set used during the training process in PyCaret.

In [ ]:
# Dictionary to map the model names to their abbreviations
model_name_to_abbr = {
    'GradientBoostingClassifier': 'Gradient Boosting Classifier',
    'RandomForestClassifier': 'Random Forest Classifier',
    'XGBClassifier': 'Extreme Gradient Boosting',
    # Add more mappings if needed
}

# Train and compare models, then select the top 3 models based on accuracy
top_3_models = compare_models(sort='Accuracy', n_select=3)

# Get the results of the model comparison in a DataFrame format
model_results = pull()

# Print the names and accuracy of the top 3 models
for i, model in enumerate(top_3_models, start=1):
    model_name = type(model).__name__
    matched_name = model_name_to_abbr.get(model_name)
    accuracy = model_results.loc[model_results['Model'] == matched_name]['Accuracy'].values[0]
    print(f"Model {i}: {model_name}, Accuracy: {accuracy:.4f}")

Processing:   0%|          | 0/67 [00:00<?, ?it/s]

Model 1: GradientBoostingClassifier, Accuracy: 0.5256
Model 2: RandomForestClassifier, Accuracy: 0.5248
Model 3: XGBClassifier, Accuracy: 0.5239


#The second result (around 52% accuracy) is the performance on the test data which was previously unseen by the model. These results show that the model is not generalizing well on the test data. Therefore, try best hyperparameter tuning and run the models again. Use tune_model() function

In [ ]:
# Initialize pycaret
exp_clf = setup(data=processed_data, target='Trend', train_size=0.8, session_id=42, preprocess=False)

# Train and compare models
models_to_compare = ['gbc', 'rf', 'xgboost'] #Top 3 performed models based on the previous code.
tuned_models = []
for model in models_to_compare:
    base_model = create_model(model)
    tuned_model = tune_model(base_model)
    tuned_models.append(tuned_model)

# Test the best tuned model on the last 100 rows of the processed_data dataframe
test_data = processed_data.iloc[-100:]
predictions = predict_model(tuned_models[0], data=test_data)

Processing:   0%|          | 0/4 [00:00<?, ?it/s]

Processing:   0%|          | 0/7 [00:00<?, ?it/s]

Fitting 10 folds for each of 10 candidates, totalling 100 fits


Original model was better than the tuned model, hence it will be returned. NOTE: The display metrics are for the tuned model (not the original one).


Processing:   0%|          | 0/4 [00:00<?, ?it/s]

Processing:   0%|          | 0/7 [00:00<?, ?it/s]

Fitting 10 folds for each of 10 candidates, totalling 100 fits


Original model was better than the tuned model, hence it will be returned. NOTE: The display metrics are for the tuned model (not the original one).


Processing:   0%|          | 0/4 [00:00<?, ?it/s]

Processing:   0%|          | 0/7 [00:00<?, ?it/s]

Fitting 10 folds for each of 10 candidates, totalling 100 fits


In [ ]:
# Get the accuracy of each model
for i, model in enumerate(tuned_models):
    model_name = str(model).split("(")[0]
    accuracy = round(model.score(processed_data.drop('Trend', axis=1), processed_data['Trend']), 4)
    print(f"{model_name}: {accuracy}")

GradientBoostingClassifier: 0.7576
RandomForestClassifier: 0.8972
XGBClassifier: 0.7085


## the accuracy values were obtained after hyperparameter tuning and using the test dataset. The accuracy values for the top 3 models (GradientBoostingClassifier, RandomForestClassifier, and XGBClassifier) were:

GradientBoostingClassifier: 0.7576

RandomForestClassifier: 0.8972

XGBClassifier: 0.7085

These accuracy values suggest that RandomForestClassifier performed the best among the three models with an accuracy of 0.8972.

##Now, we are using hyperparameter tunned RandomForestClassifier with the accuracy of 89.72% to predict the tomorrow's trend of the stock for TESLA.

#Firstly, we will save the rf model and load it later since we checked it performed the best after hyperparameter tunning.

In [ ]:
# Train and tune the RandomForestClassifier model
rf_model = tune_model(create_model('rf'))

# Save the trained and tuned RandomForestClassifier model
save_model(rf_model, 'Final RandomForest Classifier Model')

Processing:   0%|          | 0/4 [00:00<?, ?it/s]

Processing:   0%|          | 0/7 [00:00<?, ?it/s]

Fitting 10 folds for each of 10 candidates, totalling 100 fits


Original model was better than the tuned model, hence it will be returned. NOTE: The display metrics are for the tuned model (not the original one).
Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=FastMemory(location=/tmp/joblib),
          steps=[('clean_column_names',
                  TransformerWrapper(exclude=None, include=None,
                                     transformer=CleanColumnNames(match='[\\]\\[\\,\\{\\}\\"\\:]+'))),
                 ('trained_model',
                  RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                         class_weight=None, criterion='gini',
                                         max_depth=None, max_features='sqrt',
                                         max_leaf_nodes=None, max_samples=None,
                                         min_impurity_decrease=0.0,
                                         min_samples_leaf=1, min_samples_split=2,
                                         min_weight_fraction_leaf=0.0,
                                         n_estimators=100, n_jobs=-1,
                                         oob_score=False, random_state=42,
                                

In [ ]:
ticker = "TSLA"  # Tesla stock symbol

# Get the historical stock data
raw_data = get_stock_data(ticker)
processed_data = add_technical_indicators(raw_data)
processed_data = create_target_and_preprocess(processed_data)
processed_data = processed_data.dropna()

# Load the saved model
rf_model = load_model('Final RandomForest Classifier Model')

# Get the latest stock data
latest_data = get_stock_data(ticker, years=0)
latest_processed_data = add_technical_indicators(latest_data)
latest_processed_data = latest_processed_data[['Close', 'Adj Close', 'RSI', 'OBV', 'CMF', 'Fast_Stochastic_Oscillator', 'SMA', 'EMA', 'Bollinger_High', 'Bollinger_Low', 'ADX']]
latest_processed_data = latest_processed_data.iloc[-20:]

# Make predictions for tomorrow's stock trend
tomorrow_pred = predict_model(rf_model, data=latest_processed_data)
print(f"Prediction for tomorrow's stock trend: {tomorrow_pred['Label'].values[0]}")

[*********************100%***********************]  1 of 1 completed
Transformation Pipeline and Model Successfully Loaded
[*********************100%***********************]  1 of 1 completed

1 Failed download:
- TSLA: Data doesn't exist for startDate = 1681098744, endDate = 1681098744


KeyError: ignored

In [ ]:
print(processed_data)

                  Open        High         Low       Close   Adj Close  \
Date                                                                     
2017-05-09   20.625334   21.466000   20.606667   21.417334   21.417334   
2017-05-10   21.437332   21.700001   21.208000   21.681334   21.681334   
2017-05-11   21.559999   21.733334   21.306667   21.540001   21.540001   
2017-05-12   21.698668   21.799999   21.435333   21.653999   21.653999   
2017-05-15   21.225332   21.346666   20.835333   21.058666   21.058666   
...                ...         ...         ...         ...         ...   
2023-03-31  197.529999  207.789993  197.199997  207.460007  207.460007   
2023-04-03  199.910004  202.690002  192.199997  194.770004  194.770004   
2023-04-04  197.320007  198.740005  190.320007  192.580002  192.580002   
2023-04-05  190.520004  190.679993  183.759995  185.520004  185.520004   
2023-04-06  183.080002  186.389999  179.740005  185.059998  185.059998   

               Volume         %K     

In [ ]:
# Load the previously trained and tuned RandomForestClassifier model
rf_model = load_model('Final RandomForest Classifier Model')

# Make predictions for tomorrow's stock trend
tomorrow_pred = predict_model(rf_model, data=processed_data)
print(f"Prediction for tomorrow's stock trend: {tomorrow_pred['Label'].values[0]}")

Transformation Pipeline and Model Successfully Loaded


KeyError: ignored

In [ ]:
X, y = prepare_data_for_feature_selection(processed_data, 'Trend')

In [ ]:
k = 5  # Change this value to select a different number of top features
selector = SelectKBest(score_func=f_regression, k=k)
selector.fit(X, y)

cols = selector.get_support(indices=True)
selected_features = X.iloc[:, cols]
print("Top", k, "features:")
print(selected_features.columns)

Top 5 features:
Index(['Volume', '%K', 'RSI', 'ADX', 'CMF'], dtype='object')


In [ ]:
# Create a new DataFrame with the selected features, target column, and 'Close' column
final_data = pd.concat([selected_features, processed_data[['Close', 'Trend']]], axis=1)

print("\nFinal data:")
print(final_data.head())


Final data:
               Volume         %K        RSI  ADX       CMF      Close  Trend
Date                                                                        
2017-05-09  145147500  82.655837  58.706452  0.0  0.038443  21.417334      1
2017-05-10   86124000  93.387545  60.977059  0.0  0.088723  21.681334      0
2017-05-11   71307000  87.642315  59.103379  0.0  0.136940  21.540001      1
2017-05-12   61824000  92.276398  60.166585  0.0  0.115903  21.653999      0
2017-05-15  114330000  68.075857  52.491781  0.0  0.103686  21.058666      1


In [ ]:
processed_data.head(100)

Open        High         Low       Close   Adj Close  \
Date                                                                     
2020-04-24   47.387333   48.715332   46.545334   48.343334   48.343334   
2020-04-27   49.174000   53.299332   49.000000   53.250000   53.250000   
2020-04-28   53.042667   53.666668   50.445999   51.274666   51.274666   
2020-04-29   52.678001   53.546665   52.210667   53.367332   53.367332   
2020-04-30   57.012669   57.987999   50.900002   52.125332   52.125332   
...                ...         ...         ...         ...         ...   
2020-09-09  118.866669  123.000000  113.836670  122.093330  122.093330   
2020-09-10  128.736664  132.996674  120.186668  123.779999  123.779999   
2020-09-11  127.313332  127.500000  120.166664  124.239998  124.239998   
2020-09-14  126.983330  140.000000  124.433334  139.873337  139.873337   
2020-09-15  145.520004  153.979996  143.566666  149.919998  149.919998   

               Volume         %K        RSI         SMA         EMA  \
Date                                                                  
2020-04-24  198180000  82.021013  67.403192   41.086001   43.062072   
2020-04-27  310221000  99.723011  73.938099   42.033967   44.032351   
2020-04-28  228330000  86.792790  68.025545   42.923934   44.722095   
2020-04-29  243240000  98.188700  70.698777   43.845634   45.545451   
2020-04-30  427078500  69.601435  67.112503   44.846700   46.172106   
...               ...        ...        ...         ...         ...   
2020-09-09  238397400  21.087995  48.200243  132.783834  130.157518   
2020-09-10  254791800  24.019462  49.037247  133.790300  129.550135   
2020-09-11  182152500  24.818951  49.277964  134.598967  129.044408   
2020-09-14  249061800  51.990036  56.754137  136.090267  130.075734   
2020-09-15  291894600  69.451353  60.757241  137.467467  131.965664   

            Bollinger_High  Bollinger_Low        ADX         OBV       CMF  \
Date                                                                         
2020-04-24       55.379258      26.792743   0.000000  2373435000  0.016330   
2020-04-27       56.902051      27.165883   0.000000  2683656000  0.062100   
2020-04-28       57.767108      28.080760   0.000000  2455326000  0.045515   
2020-04-29       58.877755      28.813512   0.000000  2698566000  0.070513   
2020-04-30       59.272151      30.421250   0.000000  2271487500  0.041225   
...                    ...            ...        ...         ...       ...   
2020-09-09      166.046172      99.521496  39.086220  5078734800  0.176428   
2020-09-10      164.592845     102.987756  36.495480  5333526600  0.123258   
2020-09-11      163.444791     105.753143  34.087967  5515679100  0.113930   
2020-09-14      162.702161     109.478373  32.847327  5764740900  0.152542   
2020-09-15      163.948446     110.986488  32.570809  6056635500  0.115619   

            Trend  
Date               
2020-04-24      1  
2020-04-27      0  
2020-04-28      1  
2020-04-29      0  
2020-04-30      0  
...           ...  
2020-09-09      1  
2020-09-10      1  
2020-09-11      1  
2020-09-14      1  
2020-09-15      0  

[100 rows x 16 columns]

In [ ]:
processed_data.to_csv("tesla_processed_data.csv", index=False)

In [ ]:
from datetime import datetime, timedelta
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.model_selection import train_test_split

In [ ]:
X = final_data.drop('Trend', axis=1)
y = final_data['Trend']

In [ ]:
# # Prepare the dataset
# X = processed_data.drop('Trend', axis=1)
# y = processed_data['Trend']

In [ ]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=True, random_state=42)


In [ ]:
# Train the Random Forest classifier
clf = RandomForestClassifier(n_estimators=100, random_state=42)
clf.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [ ]:
# Make predictions
y_pred = clf.predict(X_test)

In [ ]:
# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)
print("Accuracy: {:.2f}%".format(accuracy * 100))
print("Classification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.5102040816326531
Accuracy: 51.02%
Classification Report:
               precision    recall  f1-score   support

           0       0.43      0.52      0.47        61
           1       0.60      0.50      0.54        86

    accuracy                           0.51       147
   macro avg       0.51      0.51      0.51       147
weighted avg       0.53      0.51      0.51       147



In [ ]:
from sklearn.linear_model import LogisticRegression

# Train the Logistic Regression classifier
clf = LogisticRegression(random_state=42)
clf.fit(X_train, y_train)

# Make predictions
y_pred = clf.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy: {:.2f}%".format(accuracy * 100))


Accuracy: 57.82%


In [ ]:
print("Target variable distribution:")
print(y.value_counts(normalize=True) * 100)

Target variable distribution:
1    53.605442
0    46.394558
Name: Trend, dtype: float64


In [ ]:
from sklearn.model_selection import TimeSeriesSplit
from sklearn.linear_model import LogisticRegression

tscv = TimeSeriesSplit(n_splits=5)

X1 = X.values
y1 = y.values

for train_index, test_index in tscv.split(X1):
    X_train1, X_test1 = X1[train_index], X1[test_index]
    y_train1, y_test1 = y1[train_index], y1[test_index]
    
    # Train the Logistic Regression classifier
    clf = LogisticRegression(random_state=42)
    clf.fit(X_train1, y_train1)

    # Make predictions
    y_pred = clf.predict(X_test1)

    # Evaluate the model
    accuracy = accuracy_score(y_test1, y_pred)
    print("Accuracy: {:.2f}%".format(accuracy * 100))

Accuracy: 46.72%
Accuracy: 56.56%
Accuracy: 53.28%
Accuracy: 59.84%
Accuracy: 46.72%


In [ ]:
# Predict tomorrow's trend
tomorrow_data = X1[-1]
tomorrow_pred = clf.predict([tomorrow_data])

# Print tomorrow's estimated trend along with the data row
print("\nTomorrow's Data:")
print(tomorrow_data)
print("\nTomorrow's Estimated Trend:", 'Uptrend' if tomorrow_pred[0] == 1 else 'Downtrend')


Tomorrow's Data:
[1.88587000e+02 1.87857383e+02 2.08640875e+02 1.68533125e+02
 1.69105047e-02 1.90410004e+02]

Tomorrow's Estimated Trend: Uptrend
